python/xi_plugin/cache.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http:#www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import bisect
+
+
+class LineCache(object):
+    """Basic access to the core's buffer.
+
+    LineCache behaves like a list of lines. Individual lines can be accessed
+    with the lines[idx] syntax. If a line is not present in the cache,
+    it will be fetched, blocking the caller until it arrives.
+    """
+    def __init__(self, total_bytes, peer, revision, test_data=None):
+        self.total_bytes = total_bytes
+        # keep ends to make calculating offsets simple
+        self.revision = revision
+        self.peer = peer
+
+        raw_data = test_data
+        if raw_data is None:
+            raw_data = peer.get_data(0, self.revision)
+        self.raw_lines = raw_data.splitlines(True) or ['']  # handle empty buffer
+        self._recalculate_offsets()
+
+    def _recalculate_offsets(self):
+        self.offsets = [0] + [len(l) for l in self.raw_lines]
+        tally = 0
+        for idx in range(len(self.offsets)):
+            new_tally = self.offsets[idx] + tally
+            self.offsets[idx] += tally
+            tally = new_tally
+
+    def __len__(self):
```

Comment:
On review:

this makes me wonder if it would make sense in certain situations to have the plugins cache metrics information as well as (and perhaps preferentially to) actual bytes. This obviously depends on use case, but it might be interesting to explore.